projects/core/src/cart/facade/cart-voucher.service.ts

```diff
@@ -0,0 +1,83 @@
+import { Injectable } from '@angular/core';
+import { select, Store } from '@ngrx/store';
+import { Observable, combineLatest } from 'rxjs';
+import { take, map } from 'rxjs/operators';
+import { AuthService } from '../../auth/index';
+import * as fromProcessStore from '../../process/store/process-state';
+import {
+  getProcessErrorFactory,
+  getProcessLoadingFactory,
+  getProcessSuccessFactory,
+} from '../../process/store/selectors/process.selectors';
+import { CartActions } from '../store/actions/index';
+import { ADD_VOUCHER_PROCESS_ID, StateWithCart } from '../store/cart-state';
+import { CartSelectors } from '../store/selectors/index';
+
+@Injectable()
+export class CartVoucherService {
+  constructor(
+    protected store: Store<
+      StateWithCart | fromProcessStore.StateWithProcess<void>
+    >,
+    protected authService: AuthService
+  ) {}
+
+  addVoucher(cartId: string, voucherId: string): void {
+    this.authService
+      .getOccUserId()
+      .pipe(take(1))
+      .subscribe(occUserId =>
+        this.store.dispatch(
+          new CartActions.CartAddVoucher({
+            userId: occUserId,
+            cartId: cartId,
+            voucherId: voucherId,
+          })
+        )
+      )
+      .unsubscribe();
+  }
+
+  removeVoucher(voucherId: string): void {
```

Comment:
What about making it consistent and more usable in other scenarios?
It would be cleaner when `addVoucher` and `removeVoucher` have similar signatures, but I got your point that we don't have `cartId` when removing voucher in our current use cases.

What do you thkink about sth like this then?
```
  addVoucher(voucherId: string, cartId?: string): void {
    this.combineUserAndCartId(cartId).subscribe(([occUserId, cartIdentifier]) =>
      this.store.dispatch(
        new CartActions.CartAddVoucher({
          userId: occUserId,
          cartId: cartIdentifier,
          voucherId: voucherId,
        })
      )
    );
  }

  removeVoucher(voucherId: string, cartId?: string): void {
    this.combineUserAndCartId(cartId).subscribe(([occUserId, cartIdentifier]) =>
      this.store.dispatch(
        new CartActions.CartRemoveVoucher({
          userId: occUserId,
          cartId: cartIdentifier,
          voucherId: voucherId,
        })
      )
    );
  }

  private combineUserAndCartId(cartId?: string): Observable<[string, string]> {
    return combineLatest([
      this.authService.getOccUserId(),
      cartId
        ? of(cartId)
        : this.store.pipe(
            select(CartSelectors.getCartContent),
            map(cart => cart.code)
          ),
    ]).pipe(take(1));
  }
```

Reply:
Done